a10_octavia/controller/worker/tasks/vthunder_tasks.py

```diff
@@ -359,3 +360,212 @@ def revert(self, vthunder, *args, **kwargs):
+        self._subnet_mask = None
+
+    @property
+    def network_driver(self):
+        if self._network_driver is None:
+            self._network_driver = a10_task_utils.get_a10_network_driver()
+        return self._network_driver
+
+    def get_tag_interface_info(self, project_id, vlan_id):
+        if project_id in CONF.hardware_thunder.devices:
+            vthunder_conf = CONF.hardware_thunder.devices[project_id]
+            if vthunder_conf.interface_vlan_map:
+                interface_vlan_map = vthunder_conf.interface_vlan_map
+                for ifnum in interface_vlan_map:
+                    vlan_dict = interface_vlan_map[ifnum]
+                    vlan_id_str = str(vlan_id)
+                    if vlan_id_str in vlan_dict:
+                        return (int(ifnum), vlan_dict[vlan_id_str])
+        return (1, None)
+
+    def get_subnet_and_mask(self, subnet_id):
+        if self._subnet is None:
+            self._subnet = self.network_driver.get_subnet(subnet_id)
+        subnet_ip, subnet_mask = a10_utils.get_net_info_from_cidr(self._subnet.cidr)
+        self._subnet_ip = subnet_ip
+        self._subnet_mask = subnet_mask
+
+    def get_vlan_id(self, subnet_id, is_revert):
+        if self._subnet is None:
+            self._subnet = self.network_driver.get_subnet(subnet_id)
+        network_id = self._subnet.network_id
+        network = self.network_driver.get_network(network_id)
+        if network.provider_network_type != 'vlan' and not is_revert:
+            raise
+        return network.provider_segmentation_id
+
+    def is_vlan_deletable(self):
+        vs_list = self.axapi_client.slb.virtual_server.get(name="")
+        server_list = self.axapi_client.slb.server.get(name="")
+
+        if vs_list:
+            for vs in vs_list["virtual-server-list"]:
+                if a10_utils.check_ip_in_subnet_range(vs["ip-address"], self._subnet_ip,
+                                                      self._subnet_mask):
+                    return False
+
+        if server_list:
+            for server in server_list["server-list"]:
+                if a10_utils.check_ip_in_subnet_range(server["host"], self._subnet_ip,
+                                                      self._subnet_mask):
+                    return False
+
+        return True
+
+    def _get_patched_ve_ip(self, ve_ip):
+        ve_ip = ve_ip.lstrip('.')
+        octets = ve_ip.split('.')
+        for idx in range(4 - len(octets)):
+            octets.insert(0, '0')
+        ve_ip = '.'.join(octets)
+        return a10_utils.merge_host_and_network_ip(self._subnet.cidr, ve_ip)
+
+    def tag_interface(self, vlan_id, ifnum, ve_info):
+        if not self.axapi_client.interface.ethernet.exists(ifnum):
+            LOG.error("ethernet interface %s does not exist", ifnum)
+            raise
+
+        eth = self.axapi_client.interface.ethernet.get(ifnum)
+        if 'action' not in eth or not eth['action'] == 'disable':
```

Comment:
Seen this code in a few places now, so I'll add the explanation here and mark with `object.get()` in the places I've seen it used. Just to make it easier to track similar code down after review.

We have syntactic sugar to handle this for us in python.

```python
if object.get('mykey', 'return_this_if_not_found') == 'a_val':
    ...
```

There's some other cool ops here: https://docs.python.org/3/library/stdtypes.html#mapping-types-dict

This specifically is for the get: https://docs.python.org/3/library/stdtypes.html#dict.get

So in this case we could do:
```python
if eth.get('action') != 'disable:
    ...
```

If no default is specified it'll just return None. So if `action` is not in `eth` then we'll have `None != disable` which is True and this will run.